File: Python/2023/day02.py
"""
--- Day 2: Cube Conundrum ---
"""

from aocd import get_data

data = get_data(day=2, year=2023).split("\n")
# Data = open("test.txt").read().split("\n")

red = 12
green = 13
blue = 14

total = 0
for line in data:
    id = line.split(":")[0].split(" ")[1]
    turns = line.split(":")[1].strip().split("; ")
    n = 0
    for t in turns:
        c = t.split(", ")
        PASSING = True
        for a in c:
            x, y = a.split(" ")
            if y == "red" and int(x) > red:
                PASSING = False
            elif y == "green" and int(x) > green:
                PASSING = False
            elif y == "blue" and int(x) > blue:
                PASSING = False
        if PASSING:
            n += 1
    if n == len(turns):
        total += int(id)

print("ONE:", total)

total = 0
for line in data:
    power = 1
    maxes = {red: 0, green: 0, blue: 0}
    id = line.split(":")[0].split(" ")[1]
    turns = line.split(":")[1].strip().split("; ")
    n = 0
    for t in turns:
        c = t.split(", ")
        for a in c:
            x, y = a.split(" ")
            if y == "red":
                maxes[red] = max(maxes[red], int(x))
            elif y == "green":
                maxes[green] = max(maxes[green], int(x))
            elif y == "blue":
                maxes[blue] = max(maxes[blue], int(x))
    for m in maxes:
        power *= maxes[m]
    total += power

print("TWO:", total)

File: C/2023/day02.c
// Inspiration: https://github.com/fuzesmarcell/aoc/blob/main/2023/day_02.c

#include <advent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Max
{
    int r;
    int g;
    int b;
} Max;

int one(FILE *fp)
{
    int  total = 0, n;
    char buf[128];

    // Get kv-pairs with colon
    while (fscanf(fp, "%s %d:", buf, &n) == 2)
    {
        const Max m = { 12, 13, 14 };
        int       i, x = 0;
        char      v[128];

        // Get kv-pairs
        while (fscanf(fp, "%d %s", &i, v) == 2)
        {
            int pass = 1;
            if (strstr(v, "red") != NULL && i > m.r)
                pass = 0;
            else if (strstr(v, "green") != NULL && i > m.g)
                pass = 0;
            else if (strstr(v, "blue") != NULL && i > m.b)
                pass = 0;

            if (!pass)
                ++x;
        }
        // X == 0 implies no failures this line
        if (x == 0)
            total += n;
    }
    return total;
}

int two(FILE *fp)
{
    int  total = 0, n;
    char buf[128];

    // Get pairs with colon after
    while (fscanf(fp, "%s %d:", buf, &n) == 2)
    {
        Max  m = { 0, 0, 0 };
        int  i;
        char v[128];
        // Get key/values
        while (fscanf(fp, "%d %s", &i, v) == 2)
        {
            // Find red/blue/green
            if (strstr(v, "red") != NULL && i > m.r)
                m.r = MAX(i, m.r);
            else if (strstr(v, "green") != NULL && i > m.g)
                m.g = MAX(i, m.g);
            else if (strstr(v, "blue") != NULL && i > m.b)
                m.b = MAX(i, m.b);
        }
        total += (m.r * m.g * m.b);
    }
    return total;
}

int main(void)
{
    FILE *fp = fopen("day02.txt", "r");
    if (!fp)
    {
        printf("Bad file read\n");
        exit(EXIT_FAILURE);
    }

    printf("ONE: %d\n", one(fp));
    fseek(fp, 0, 0);
    printf("TWO: %d\n", two(fp));

    fclose(fp);
    return 0;
}
